### PR TITLE
migrate to typescript

### PR DESCRIPTION
**Pull request by Joystick01**

### CODE REVIEW

File: src/helpers/cosmosDbHelper.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    public async deleteMessages(userID: string): Promise<void> {
+        const hashedUserID = hashUserID.hashUserID(userID);
+        if (await this.userExists(userID)) {
+            await this.container.item(hashedUserID, hashedUserID).replace({ id: hashedUserID, messages: [] });
+        }
+    }
+
+}
+
+class hashUserID {
```
Comment (Uggah) [inline src/helpers/cosmosDbHelper.ts:59]: Class names should be Upper-Camel-Case. Besides that, it is considered bad practice to have two classes in one file.

File: src/helpers/ingressHandler.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+    const openAiHelper = new OpenAiHelper(process.env["OPENAI_ENDPOINT"], process.env["OPENAI_KEY"], process.env["OPENAI_VERSION"], process.env["OPENAI_DEPLOYMENT"], process.env["OPENAI_SYSTEM_PROMPT"], parseInt(process.env["OPENAI_MAX_TOKENS"]));
+    const response = await openAiHelper.generateResponse(messages);
+    
+    cosmosDbHelper.saveMessage(userID, response, "assistant");
+    smsEgressHandler.sendSMS(userID, response);
+
+}
+
+type UserProperties = {
```
Comment (Uggah) [inline src/helpers/ingressHandler.ts:29]: Why not put the type in a separate file?

File: src/helpers/cosmosDbHelper.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    private database: Database;
+    private container: Container;
+
+    constructor(private readonly connectionString: string, private readonly databaseName: string, private readonly containerName: string) {
+        this.client = new CosmosClient(connectionString);
+        this.database = this.client.database(databaseName);
+        this.container = this.database.container(containerName);
+    }
+
+    async userExists(userID: string): Promise<boolean>  {
```
Comment (Uggah) [inline src/helpers/cosmosDbHelper.ts:17]: private?

File: src/helpers/cosmosDbHelper.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
+import { Container, CosmosClient, Database, ResourceResponse } from "@azure/cosmos";
+import { createHash } from "crypto";
+import { ChatCompletionMessageParam, ChatCompletionRole } from "openai/resources/chat/completions";
+
+export class CosmosDbHelper {
+
+    private client: CosmosClient;
```
Comment (Uggah) [inline src/helpers/cosmosDbHelper.ts:7]: Why not `private readonly`?

File: src/helpers/ingressHandler.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import { InvocationContext } from '@azure/functions';
+import { CosmosDbHelper } from './cosmosDbHelper';
+import { OpenAiHelper } from './openAiHelper';
+import { SmsEgressHandler } from './smsEgressHandler';
+
+export async function ingressHandler(message: string, context: InvocationContext): Promise<void> {
+
+    const userID = (context.triggerMetadata.userProperties as UserProperties).address;
```
Comment (Uggah) [inline src/helpers/ingressHandler.ts:8]: I like userId more :P - if we stick with userID throughout the code, I will let this one slip :laughing:  :+1:

File: src/helpers/ingressHandler.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    const cosmosDbHelper = new CosmosDbHelper(process.env["COSMOSDB_CONNECTION"], process.env["COSMOSDB_DATABASE"], process.env["COSMOSDB_CONTAINER"]);
+    const smsEgressHandler = new SmsEgressHandler(process.env["SERVICEBUS_CONNECTION_EGRESS"], process.env["EGRESS_QUEUE"]);
+
+    if (message === "STOP") {
+        cosmosDbHelper.deleteMessages(userID);
+        smsEgressHandler.sendSMS(userID, "<Chat has been reset>\n\nHi I'm Guppy-AI how can I help you today?");
+        return;
+    }
+
+    const messages = await cosmosDbHelper.saveMessage(userID, message, "user");
```
Comment (Uggah) [inline src/helpers/ingressHandler.ts:19]: Very much optional and highly subjective, but I prefer string-based enums over hard-coded strings in code.

File: src/helpers/openAiHelper.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import { AzureOpenAI } from "openai";
+import { ChatCompletionMessageParam } from "openai/resources";
+
+export class OpenAiHelper {
+
+    private client: AzureOpenAI;
+    private maxTokens: number;
+
+    constructor(private readonly endpoint: string, private readonly apiKey: string, private readonly apiVersion: string, private readonly deployment: string, private readonly system_prompt: string,  maxTokens: number) {
```
Comment (Uggah) [inline src/helpers/openAiHelper.ts:9]: system_prompt -> systemPrompt

File: src/helpers/openAiHelper.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
+import { AzureOpenAI } from "openai";
+import { ChatCompletionMessageParam } from "openai/resources";
+
+export class OpenAiHelper {
+
+    private client: AzureOpenAI;
```
Comment (Uggah) [inline src/helpers/openAiHelper.ts:6]: `private readonly`

File: src/helpers/smsEgressHandler.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
+import { ServiceBusClient, ServiceBusSender } from "@azure/service-bus";
+
+export class SmsEgressHandler {
+    private client: ServiceBusClient
```
Comment (Uggah) [inline src/helpers/smsEgressHandler.ts:4]: Do we need the client after the sender has been created?

File: src/helpers/smsEgressHandler.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+    private sender: ServiceBusSender;
+    private static utf8Encode = new TextEncoder();
```
Comment (Uggah) [inline src/helpers/smsEgressHandler.ts:6]: Elegant :+1:

**Review by Uggah — COMMENTED** [review]
Some thoughts about the code...

File: src/helpers/ingressHandler.ts (hunk 6 above)
Comment (Joystick01) [reply src/helpers/ingressHandler.ts:19]: this is semi-forced by the OpenAI library

**Pull request merged**